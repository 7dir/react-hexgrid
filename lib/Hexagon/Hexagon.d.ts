import * as React from "react";
import { Hex } from "../models/Hex";
import { Point } from "../models/Point";
type H = {
    data?: any;
    state: {
        hex: Hex;
    };
    props: HexagonProps;
};
export type HexagonDragEventHandler<T = Element, AdditionalData = any> = (event: React.DragEvent<T>, h: H, additionalData?: AdditionalData) => void;
export type HexagonDragDropEventHandler<T = Element, AdditionalData = any> = (event: React.DragEvent<T>, h: H, additionalData: AdditionalData) => void;
export type HexagonMouseEventHandler<T = SVGGElement> = (event: React.MouseEvent<T, globalThis.MouseEvent>, h: H) => void;
export type HexagonProps = {
    /**
     * The q coordinate of the hexagon.
     */
    q: number;
    /**
     * The r coordinate of the hexagon.
     */
    r: number;
    /**
     * The s coordinate of the hexagon.
     */
    s: number;
    /**
     * The pattern id for the fill image of the hexagon
     */
    fill?: string;
    /**
     * Optional classname.  Usage unclear, currently not used.
     */
    className?: string;
    /**
     * The circumradius (from origin to vertex) of the hexagon.  Defaults to 10.
     */
    radius?: number;
    /**
     * CSSProperties for the hexagon.  cellStyle is deprecated and only remains to avoid breaking changes
     */
    cellStyle?: React.CSSProperties | undefined;
    style?: React.CSSProperties | undefined;
    data?: any;
    onMouseEnter?: HexagonMouseEventHandler;
    onMouseLeave?: HexagonMouseEventHandler;
    onClick?: HexagonMouseEventHandler;
    onDragStart?: HexagonDragEventHandler;
    onDragEnd?: HexagonDragEventHandler;
    onDragOver?: HexagonDragEventHandler;
    onDrop?: HexagonDragDropEventHandler<any, TargetProps>;
    onMouseOver?: HexagonMouseEventHandler;
    children?: React.ReactNode | React.ReactNode[];
};
type TargetProps = {
    hex: Hex;
    pixel: Point;
    data?: any;
    fill?: string;
    className?: string;
};
/**
 * Renders a Hexagon cell at the given qrs-based coordinates.
 */
export declare function Hexagon(props: HexagonProps & Omit<React.SVGProps<SVGGElement>, "transform" | "onDragStart" | "onDragEnd" | "onDrop" | "onDragOver" | "onMouseEnter" | "onClick" | "onMouseOver" | "onMouseLeave">): React.JSX.Element;
export default Hexagon;
//# sourceMappingURL=Hexagon.d.ts.map