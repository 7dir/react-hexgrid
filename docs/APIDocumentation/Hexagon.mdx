import {ArgTypes, Meta} from '@storybook/blocks';

import * as HexagonComponentStory from '../../src/Hexagon/Hexagon.stories.tsx';

<Meta title="API Documentation/Hexagon"/>

# Hexagon 

<ArgTypes of={HexagonComponentStory.Primary}/>