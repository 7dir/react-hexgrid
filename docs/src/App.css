@font-face {
    font-family: 'SauceCodePro Nerd Font';
    src: local("./SauceCodeProNerdFontMono-Black.ttf");
    format: ('truetype');
}

* {
    font-family: "SauceCodePro Nerd Font", serif;
}