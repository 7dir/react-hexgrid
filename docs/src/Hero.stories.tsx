import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import Hero from "./Hero"

const meta = {
    title: "Stories/Hero",
} satisfies Meta<typeof Hero>

export default meta;

type Story = StoryObj<typeof Hero>

export const Default: Story = {
    render: () => (
        <Hero />
    )
}

