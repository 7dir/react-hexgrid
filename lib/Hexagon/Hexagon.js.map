{"version":3,"file":"Hexagon.js","sourceRoot":"","sources":["../../src/Hexagon/Hexagon.tsx"],"names":[],"mappings":";;;;;;;;;;;AAAA,OAAO,KAAK,KAAK,MAAM,OAAO,CAAA;AAC9B,qCAAqC;AACrC,OAAO,EAAE,GAAG,EAAE,MAAM,eAAe,CAAA;AACnC,OAAO,EAAE,QAAQ,EAAE,MAAM,aAAa,CAAA;AACtC,OAAO,EAAmB,gBAAgB,EAAE,MAAM,WAAW,CAAA;AAC7D,OAAO,EAAE,KAAK,EAAE,MAAM,iBAAiB,CAAA;AACvC,OAAO,EAAE,WAAW,EAAE,MAAM,WAAW,CAAC;AACxC,OAAO,EAAE,oBAAoB,EAAE,MAAM,WAAW,CAAA;AA0EhD;;GAEG;AACH,MAAM,UAAU,OAAO,CACrB,KAYG;IAEH,sCAAsC;IACtC,MAAM,EACJ,CAAC,EACD,CAAC,EACD,CAAC,EACD,IAAI,EACJ,SAAS,EACT,KAAK,EACL,SAAS,EACT,QAAQ,EACR,WAAW,EACX,SAAS,EACT,MAAM,EACN,UAAU,EACV,YAAY,EACZ,YAAY,EACZ,WAAW,EACX,OAAO,EACP,IAAI,EACJ,WAAW,KAET,KAAK,EADJ,IAAI,UACL,KAAK;IAET,gDAAgD;IAEhD;;;MAGE;IAEF;;;;OAIG;MAjCG,2MAoBL,CAAQ,CAAA;IAET,gDAAgD;IAEhD;;;MAGE;IAEF;;;;OAIG;IACH,MAAM,WAAW,GAAG,IAAI,KAAK,CAAC,EAAE,EAAE,EAAE,CAAC,CAAA;IACrC,MAAM,aAAa,GAAG,IAAI,KAAK,CAAC,CAAC,EAAE,CAAC,CAAC,CAAA;IACrC,MAAM,cAAc,GAAG,GAAG,CAAA;IAE1B,MAAM,iBAAiB,GAAoB,CAAC;QAC1C,IAAI,EAAE,WAAW;QACjB,WAAW,EAAE,WAAW;QACxB,MAAM,EAAE,aAAa;QACrB,OAAO,EAAE,cAAc;KACxB,CAAC,CAAA;IAEF,6FAA6F;IAC7F,MAAM,MAAM,GAAG,gBAAgB,EAAE,CAAC,MAAM,IAAI,iBAAiB,CAAA;IAC7D,MAAM,MAAM,GAAG,gBAAgB,EAAE,CAAC,MAAM,IAAI,oBAAoB,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC,CAAC,CAAC,GAAG,CAAC,CAAC,KAAK,EAAE,EAAE,CAAC,GAAG,KAAK,CAAC,CAAC,IAAI,KAAK,CAAC,CAAC,EAAE,CAAC,CAAC,IAAI,CAAC,GAAG,CAAC,CAAC;IAGlI,MAAM,EAAE,GAAG,EAAE,KAAK,EAAE,GAAG,KAAK,CAAC,OAAO,CAAC,GAAG,EAAE;QACxC,MAAM,GAAG,GAAG,IAAI,GAAG,CAAC,CAAC,EAAE,CAAC,EAAE,CAAC,CAAC,CAAA;QAC5B,MAAM,KAAK,GAAG,QAAQ,CAAC,UAAU,CAAC,GAAG,EAAE,MAAM,CAAC,CAAA;QAC9C,OAAO;YACL,GAAG;YACH,KAAK;SACN,CAAA;IACH,CAAC,EAAE,CAAC,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,MAAM,CAAC,CAAC,CAAA;IAErB,8BAA8B;IAC9B,MAAM,KAAK,GAAG,EAAE,GAAG,EAAE,CAAA;IAErB,4FAA4F;IAC5F,+DAA+D;IAC/D,uEAAuE;IACvE,MAAM,MAAM,GAAG,IAAI,CAAC,CAAC,CAAC,QAAQ,IAAI,GAAG,CAAC,CAAC,CAAC,SAAS,CAAA;IACjD,MAAM,SAAS,GAAG,EAAE,SAAS,EAAE,IAAI,EAAS,CAAA;IAC5C,OAAO,CACL,yCACE,SAAS,EAAE,SAAS,EACpB,SAAS,EAAE,aAAa,KAAK,CAAC,CAAC,KAAK,KAAK,CAAC,CAAC,GAAG,IAC1C,IAAI,EACJ,SAAS,IACb,WAAW,EAAE,CAAC,CAAC,EAAE,EAAE;YACjB,IAAI,WAAW,EAAE,CAAC;gBAChB,MAAM,WAAW,GAAgB;oBAC/B,GAAG,EAAE,GAAG;oBACR,KAAK;oBACL,IAAI,EAAE,IAAI;oBACV,IAAI,EAAE,IAAI;oBACV,SAAS,EAAE,SAAS;iBACrB,CAAA;gBACD,CAAC,CAAC,YAAY,CAAC,OAAO,CAAC,SAAS,EAAE,IAAI,CAAC,SAAS,CAAC,WAAW,CAAC,CAAC,CAAA;gBAC9D,WAAW,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACxC,CAAC;QACH,CAAC,EACD,SAAS,EAAE,CAAC,CAAC,EAAE,EAAE;YACf,IAAI,SAAS,EAAE,CAAC;gBACd,CAAC,CAAC,cAAc,EAAE,CAAA;gBAClB,MAAM,OAAO,GAAG,CAAC,CAAC,YAAY,CAAC,UAAU,KAAK,MAAM,CAAA;gBACpD,SAAS,CAAC,CAAC,EAAE,EAAE,KAAK,EAAE,KAAK,EAAE,EAAE,OAAO,CAAC,CAAA;YACzC,CAAC;QACH,CAAC,EACD,MAAM,EAAE,CAAC,CAAC,EAAE,EAAE;YACZ,IAAI,MAAM,EAAE,CAAC;gBACX,CAAC,CAAC,cAAc,EAAE,CAAA;gBAClB,MAAM,MAAM,GAAG,IAAI,CAAC,KAAK,CAAC,CAAC,CAAC,YAAY,CAAC,OAAO,CAAC,SAAS,CAAC,CAAC,CAAA;gBAC5D,MAAM,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,EAAE,MAAM,CAAC,CAAA;YAC3C,CAAC;QACH,CAAC,EACD,UAAU,EAAE,CAAC,CAAC,EAAE,EAAE;YAChB,IAAI,UAAU,EAAE,CAAC;gBACf,UAAU,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACvC,CAAC;QACH,CAAC,EACD,YAAY,EAAE,CAAC,CAAC,EAAE,EAAE;YAClB,IAAI,YAAY,EAAE,CAAC;gBACjB,YAAY,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACzC,CAAC;QACH,CAAC,EACD,OAAO,EAAE,CAAC,CAAC,EAAE,EAAE;YACb,IAAI,OAAO,EAAE,CAAC;gBACZ,OAAO,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACpC,CAAC;QACH,CAAC,EACD,WAAW,EAAE,CAAC,CAAC,EAAE,EAAE;YACjB,IAAI,WAAW,EAAE,CAAC;gBAChB,WAAW,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACxC,CAAC;QACH,CAAC,EACD,YAAY,EAAE,CAAC,CAAC,EAAE,EAAE;YAClB,IAAI,YAAY,EAAE,CAAC;gBACjB,YAAY,CAAC,CAAC,EAAE,EAAE,IAAI,EAAE,KAAK,EAAE,KAAK,EAAE,CAAC,CAAA;YACzC,CAAC;QACH,CAAC;QAED,2BAAG,SAAS,EAAC,SAAS;YACpB,iCAAS,MAAM,EAAE,MAAM,EAAE,IAAI,EAAE,MAAM,EAAE,KAAK,EAAE,SAAS,IAAI,KAAK,GAAI;YACnE,QAAQ,CACP,CACF,CACL,CAAA;AACH,CAAC;AAED,eAAe,OAAO,CAAA"}