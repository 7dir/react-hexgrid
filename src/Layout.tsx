import * as React from "react"
import { Orientation } from "./models/Orientation"
import { Point } from "./models/Point"

export type Size = { x: number; y: number }

export type LayoutDimension = {
  size: Size
  orientation: Orientation
  origin: Size
  spacing: number
}
export type LayoutContextProps = {
  layout: LayoutDimension
  points: string
}

const LAYOUT_FLAT = new Orientation(
  3.0 / 2.0,
  0.0,
  Math.sqrt(3.0) / 2.0,
  Math.sqrt(3.0),
  2.0 / 3.0,
  0.0,
  -1.0 / 3.0,
  Math.sqrt(3.0) / 3.0,
  0.0,
)
const LAYOUT_POINTY = new Orientation(
  Math.sqrt(3.0),
  Math.sqrt(3.0) / 2.0,
  0.0,
  3.0 / 2.0,
  Math.sqrt(3.0) / 3.0,
  -1.0 / 3.0,
  0.0,
  2.0 / 3.0,
  0.5,
)
const defaultSize = new Point(10, 10)
const defaultOrigin = new Point(0, 0)
const defaultSpacing = 1.0

const Context = React.createContext<LayoutContextProps>({
  layout: {
    size: defaultSize,
    orientation: LAYOUT_FLAT,
    origin: defaultOrigin,
    spacing: defaultSpacing,
  },
  points: "",
})

export function useLayoutContext() {
  const ctx = React.useContext(Context)
  return ctx
}

function getPointOffset(corner, orientation: Orientation, size: Size) {
  let angle = (2.0 * Math.PI * (corner + orientation.startAngle)) / 6
  return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle))
}

// TODO Refactor
function calculateCoordinates(orientation: Orientation, size: Size) {
  const corners = []
  const center = new Point(0, 0)

  Array.from(new Array(6), (x, i) => {
    const offset = getPointOffset(i, orientation, size)
    const point = new Point(center.x + offset.x, center.y + offset.y)
    corners.push(point)
  })

  return corners
}

export type LayoutProps = {
  children:
    | React.ReactElement
    | React.ReactElement[]
    | JSX.Element
    | JSX.Element[]
  className?: string
  flat?: boolean
  origin?: any
  size?: Size
  spacing?: number
}

export function Layout({
  size = defaultSize,
  flat = true,
  spacing = defaultSpacing,
  origin = defaultOrigin,
  children,
  className,
  ...rest
}: LayoutProps) {
  const orientation = flat ? LAYOUT_FLAT : LAYOUT_POINTY
  const cornerCoords = calculateCoordinates(orientation, size)
  const points = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ")
  const childLayout = Object.assign({}, rest, {
    orientation,
    size,
    origin,
    spacing,
  })

  // console.log({ childLayout })

  return (
    <Context.Provider
      value={{
        layout: childLayout,
        points,
      }}
    >
      <g className={className}>{children}</g>
    </Context.Provider>
  )
}

export default Layout
