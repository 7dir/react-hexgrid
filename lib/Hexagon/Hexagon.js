var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from "react";
//import classNames from "classnames"
import { Hex } from "../models/Hex";
import { HexUtils } from "../HexUtils";
import { useLayoutContext } from "../Layout";
import { Point } from "../models/Point";
import { LAYOUT_FLAT } from "../Layout";
import { calculateCoordinates } from "../Layout";
/**
 * Renders a Hexagon cell at the given qrs-based coordinates.
 */
export function Hexagon(props) {
    // destructure props into their values
    const { q, r, s, fill, cellStyle, style, className, radius, children, onDragStart, onDragEnd, onDrop, onDragOver, onMouseEnter, onMouseLeave, onMouseOver, onClick, data, fillOpacity } = props, rest = __rest(props
    // const { layout, points } = useLayoutContext()
    /*
    from ../Layout.tsx
    
    */
    /**
     * Hexagons need default size, origin, and spacing values to be rendered.
     * These are the default values used if no layout is provided.
     * These are copied from Layout.tsx
     */
    , ["q", "r", "s", "fill", "cellStyle", "style", "className", "radius", "children", "onDragStart", "onDragEnd", "onDrop", "onDragOver", "onMouseEnter", "onMouseLeave", "onMouseOver", "onClick", "data", "fillOpacity"]);
    // const { layout, points } = useLayoutContext()
    /*
    from ../Layout.tsx
    
    */
    /**
     * Hexagons need default size, origin, and spacing values to be rendered.
     * These are the default values used if no layout is provided.
     * These are copied from Layout.tsx
     */
    const defaultSize = new Point(10, 10);
    const defaultOrigin = new Point(0, 0);
    const defaultSpacing = 1.0;
    const flatDefaultLayout = ({
        size: defaultSize,
        orientation: LAYOUT_FLAT,
        origin: defaultOrigin,
        spacing: defaultSpacing,
    });
    // Here we look for a layout context (layout parent).  If it isn't there, we use the default.
    const layout = useLayoutContext().layout || flatDefaultLayout;
    // if radius is set, we use that, otherwise we know layout is guaranteed so we can safely use that
    const smartRadius = radius || layout.size.x;
    const points = useLayoutContext().points || calculateCoordinates(smartRadius).map((point) => `${point.x},${point.y}`).join(" ");
    const { hex, pixel } = React.useMemo(() => {
        const hex = new Hex(q, r, s);
        const pixel = HexUtils.hexToPixel(hex, layout);
        return {
            hex,
            pixel,
        };
    }, [q, r, s, layout]);
    // for backwards comapatbility
    const state = { hex };
    // the fill prop used to be set to fillId.  However, it doesn't work and throughout the code
    // we just use css with the g or polygon selectors to set fill.
    // I'm actually not sure what this prop and variable are supposed to do
    const fillId = fill ? `url(#${fill})` : undefined;
    const draggable = { draggable: true };
    return (React.createElement("g", Object.assign({ className: className, transform: `translate(${pixel.x}, ${pixel.y})` }, rest, draggable, { onDragStart: (e) => {
            if (onDragStart) {
                const targetProps = {
                    hex: hex,
                    pixel,
                    data: data,
                    fill: fill,
                    className: className,
                };
                e.dataTransfer.setData("hexagon", JSON.stringify(targetProps));
                onDragStart(e, { data, state, props });
            }
        }, onDragEnd: (e) => {
            if (onDragEnd) {
                e.preventDefault();
                const success = e.dataTransfer.dropEffect !== "none";
                onDragEnd(e, { state, props }, success);
            }
        }, onDrop: (e) => {
            if (onDrop) {
                e.preventDefault();
                const target = JSON.parse(e.dataTransfer.getData("hexagon"));
                onDrop(e, { data, state, props }, target);
            }
        }, onDragOver: (e) => {
            if (onDragOver) {
                onDragOver(e, { data, state, props });
            }
        }, onMouseEnter: (e) => {
            if (onMouseEnter) {
                onMouseEnter(e, { data, state, props });
            }
        }, onClick: (e) => {
            if (onClick) {
                onClick(e, { data, state, props });
            }
        }, onMouseOver: (e) => {
            if (onMouseOver) {
                onMouseOver(e, { data, state, props });
            }
        }, onMouseLeave: (e) => {
            if (onMouseLeave) {
                onMouseLeave(e, { data, state, props });
            }
        } }),
        React.createElement("g", { className: "hexagon" },
            React.createElement("polygon", { points: points, fill: fillId, style: cellStyle || style }),
            children)));
}
export default Hexagon;
//# sourceMappingURL=Hexagon.js.map