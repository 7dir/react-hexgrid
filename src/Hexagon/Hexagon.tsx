import * as React from "react"
import classNames from "classnames"
import { Hex } from "../models/Hex"
import HexUtils from "../HexUtils"
import { useLayoutContext } from "../Layout"

type H = { data?: any; state: { hex: Hex }; props?: { fill?: string } }

export type HexagonDragEventHandler<T = Element, AdditionalData = any> = (
  event: React.DragEvent<T>,
  h: H,
  additionalData?: AdditionalData,
) => void

export type HexagonDragDropEventHandler<T = Element, AdditionalData = any> = (
  event: React.DragEvent<T>,
  h: H,
  additionalData: AdditionalData,
) => void

export type HexagonMouseEventHandler<T = SVGGElement> = (
  event: React.MouseEvent<T, globalThis.MouseEvent>,
  h: H,
) => void

export type HexagonProps = {
  q: number
  r: number
  s: number
  fill?: string
  className?: string
  cellStyle?: React.CSSProperties | undefined
  data?: any
  onDragStart?: HexagonDragEventHandler
  onDragEnd?: HexagonDragEventHandler
  onDrop?: HexagonDragDropEventHandler<any, TargetProps>
  onDragOver?: HexagonDragEventHandler
  onMouseEnter?: HexagonMouseEventHandler
  onMouseLeave?: HexagonMouseEventHandler
  onClick?: HexagonMouseEventHandler
  onMouseOver?: HexagonMouseEventHandler
  children?: React.ReactNode | React.ReactNode[]
}

type TargetProps = {
  hex: Hex
  data?: any
  fill?: string
  className?: string
}

export function Hexagon({
  q,
  r,
  s,
  fill,
  cellStyle,
  className,
  children,
  onDragStart,
  onDragEnd,
  onDrop,
  onDragOver,
  onMouseEnter,
  onMouseLeave,
  onMouseOver,
  onClick,
  data,
  fillOpacity,
  ...rest
}: HexagonProps &
  Omit<
    React.SVGProps<SVGGElement>,
    | "transform"
    | "onDragStart"
    | "onDragEnd"
    | "onDrop"
    | "onDragOver"
    | "onMouseEnter"
    | "onClick"
    | "onMouseOver"
    | "onMouseLeave"
  >) {
  const { layout, points } = useLayoutContext()

  const { hex, pixel } = React.useMemo(() => {
    const hex = new Hex(q, r, s)
    // console.log({ hex, layout })
    const pixel = HexUtils.hexToPixel(hex, layout)
    return {
      hex,
      pixel,
    }
  }, [q, r, s, layout])

  // for backwards comapatbility
  const state = { hex }

  const fillId = fill ? `url(#${fill})` : undefined
  const draggable = { draggable: true } as any
  return (
    <g
      className={classNames("hexagon-group", className)}
      transform={`translate(${pixel.x}, ${pixel.y})`}
      // {...rest}
      {...draggable}
      onDragStart={(e) => {
        if (onDragStart) {
          const targetProps: TargetProps = {
            hex: hex,
            // state: { hex, pixel },
            // ...{ hex, pixel },
            // ...this.state,
            data: data,
            fill: fill,
            className: className,
          }
          e.dataTransfer.setData("hexagon", JSON.stringify(targetProps))
          onDragStart(e, { data, state })
        }
      }}
      onDragEnd={(e) => {
        if (onDragEnd) {
          e.preventDefault()
          const success = e.dataTransfer.dropEffect !== "none"
          onDragEnd(e, { state }, success)
        }
      }}
      onDrop={(e) => {
        if (onDrop) {
          e.preventDefault()
          const target = JSON.parse(e.dataTransfer.getData("hexagon"))
          onDrop(e, { data, state }, target)
        }
      }}
      onDragOver={(e) => {
        if (onDragOver) {
          onDragOver(e, { data, state })
        }
      }}
      onMouseEnter={(e) => {
        if (onMouseEnter) {
          onMouseEnter(e, { data, state })
        }
      }}
      // onMouseOver={(e) => this.onMouseOver(e)}
      onClick={(e) => {
        if (onClick) {
          onClick(e, { data, state })
        }
      }}
      onMouseOver={(e) => {
        if (onMouseOver) {
          onMouseOver(e, { data, state })
        }
      }}
      onMouseLeave={(e) => {
        if (onMouseLeave) {
          onMouseLeave(e, { data, state })
        }
      }}

      // onDragStart={(e) => this.onDragStart(e)}
      // onDragEnd={(e) => this.onDragEnd(e)}
      // onDragOver={(e) => this.onDragOver(e)}
      // onDrop={(e) => this.onDrop(e)}
    >
      <g className="hexagon">
        <polygon points={points} fill={fillId} style={cellStyle} />
        {children}
      </g>
    </g>
  )
}

export default Hexagon
