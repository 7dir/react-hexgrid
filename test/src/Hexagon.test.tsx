import React from "react"
import { render, fireEvent } from "@testing-library/react"

import { Layout } from "../../src/Layout"
import { Hexagon } from "../../src/Hexagon/Hexagon"

test("Hexagon should render correctly with default props", () => {
  const { container } = render(
    <svg>
      <Layout
        className={"test1"}
        size={{ x: 6, y: 6 }}
        flat={false}
        spacing={1.1}
        origin={{ x: 0, y: 0 }}
      >
        <Hexagon q={0} r={0} s={0}>
          <div>child</div>
        </Hexagon>
      </Layout>
    </svg>,
  )

  expect(container.firstChild).toMatchSnapshot()
})

test("Hexagon mouse callbacks should be called", async () => {
  const onMouseEnter = jest.fn()
  const onMouseOver = jest.fn()
  const onMouseLeave = jest.fn()
  const onClick = jest.fn()
  const onDragStart = jest.fn()
  const onDragEnd = jest.fn()
  const onDragOver = jest.fn()
  const onDrop = jest.fn()

  const { container } = render(
    <svg>
      <Layout
        className={"layout"}
        size={{ x: 6, y: 6 }}
        flat={false}
        spacing={1.1}
        origin={{ x: 0, y: 0 }}
      >
        <Hexagon
          q={0}
          r={0}
          s={0}
          fill={"#333"}
          className={"test1"}
          data={{ a: "b" }}
          onMouseEnter={onMouseEnter}
          onMouseOver={onMouseOver}
          onMouseLeave={onMouseLeave}
          onClick={onClick}
          onDragStart={onDragStart}
          onDragEnd={onDragEnd}
          onDragOver={onDragOver}
          onDrop={onDrop}
        >
          <div>child</div>
        </Hexagon>
      </Layout>
    </svg>,
  )

  const el = container.getElementsByClassName("test1")[0]

  expect(el).toBeDefined()

  expect(onMouseOver).toHaveBeenCalledTimes(0)
  expect(onMouseEnter).toHaveBeenCalledTimes(0)
  fireEvent.mouseEnter(el)
  expect(onMouseEnter).toHaveBeenCalledTimes(1)
  // mouse over seems to be also called on mouse enter
  expect(onMouseOver).toHaveBeenCalledTimes(1)

  fireEvent.mouseOver(el)
  expect(onMouseOver).toHaveBeenCalledTimes(2)

  fireEvent.mouseLeave(el)
  expect(onMouseLeave).toHaveBeenCalledTimes(1)

  fireEvent.click(el)
  expect(onClick).toHaveBeenCalledTimes(1)

  fireEvent.dragStart(el, { dataTransfer: { setData: () => {} } })

  expect(onDragStart).toHaveBeenCalledTimes(1)

  fireEvent.dragEnd(el, { dataTransfer: { setData: () => {} } })

  expect(onDragEnd).toHaveBeenCalledTimes(1)

  fireEvent.dragOver(el)
  expect(onDragOver).toHaveBeenCalledTimes(1)

  fireEvent.drop(el, {
    dataTransfer: { getData: (data) => JSON.stringify({ data }) },
  })
  expect(onDrop).toHaveBeenCalledTimes(1)
})
