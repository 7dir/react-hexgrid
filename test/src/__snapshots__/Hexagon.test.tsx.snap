// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hexagon should render correctly with default props 1`] = `
<svg>
  <g
    class="test1"
  >
    <g
      class="hexagon-group"
      draggable="true"
      transform="translate(0, 0)"
    >
      <g
        class="hexagon"
      >
        <polygon
          points="5.196152422706632,2.9999999999999996 3.6739403974420594e-16,6 -5.196152422706631,3.0000000000000018 -5.196152422706633,-2.9999999999999982 -1.102182119232618e-15,-6 5.196152422706633,-2.9999999999999982"
        />
        <div>
          child
        </div>
      </g>
    </g>
  </g>
</svg>
`;
